ユーザー管理機能

tamtam-time opened the pull request:
https://gyazo.com/d5375ba1ff21a0f30745a2ecf013905c

https://gyazo.com/69055dc7852f2831ea1e844c326d8671

https://gyazo.com/00fb80d4dda99c2ae4b28163f9a41945

https://gyazo.com/09b7ed69ac963ab31458a0a81bd0da2a

https://gyazo.com/e3a87ebaef313cf9b8cf5c6cf9450f94

https://gyazo.com/365fad8ffd5bb4984dd0dbfb47a3c168

tamtam-time committed:
ユーザー管理機能

https://gyazo.com/d5375ba1ff21a0f30745a2ecf013905c

https://gyazo.com/69055dc7852f2831ea1e844c326d8671

https://gyazo.com/00fb80d4dda99c2ae4b28163f9a41945

https://gyazo.com/09b7ed69ac963ab31458a0a81bd0da2a

https://gyazo.com/e3a87ebaef313cf9b8cf5c6cf9450f94

https://gyazo.com/365fad8ffd5bb4984dd0dbfb47a3c168

tamtam-time commented:
# What
ユーザー管理機能の実装

# Why
ユーザー管理の為

ShotaOtake commented on app/controllers/items_controller.rb:
```diff
@@ -1,2 +1,20 @@
 class ItemsController < ApplicationController
+  before_action :move_to_index, except: [:index, :show]
```
◎内容をコメントアウトしましょう

```
before_action :move_to_index, except: [:index, :show]
```

ユーザー管理機能以外の他機能の内容についてはすべて内容を削除もしくはコメントアウトしましょう。

機能実装時に差分として反映させるためです。

ShotaOtake commented on app/models/user.rb:
```diff
@@ -0,0 +1,22 @@
+  devise :database_authenticatable, :registerable,
+         :recoverable, :rememberable, :validatable
+         
+  has_many :items
```
◎内容をコメントアウトしましょう

```
  has_many :items
  has_many :historys
```

ユーザー管理機能以外の他機能の内容についてはすべて内容を削除もしくはコメントアウトしましょう。

機能実装時に差分として反映させるためです。

ShotaOtake commented on app/models/user.rb:
```diff
@@ -0,0 +1,22 @@
+  validates :email, presence: true
+  validates :password, presence: true, format: { with: /\A[a-z0-9]+\z/, message: "半角英数字で入力してください" },length: { minimum: 6 }
+  validates :password_confirmation, presence: true, format: { with: /\A[a-z0-9]+\z/, message: "半角英数字で入力してください" },length: { minimum: 6 }
+  validates :first_name, presence: true, format: { with: /\A[ぁ-んァ-ン一-龥]/, message: "全角で入力してください" }
```
◎正規表現を見直しましょう！

```
with_options format: { with: /\A[ぁ-んァ-ン一-龥]/ } do
```

現状、『々』や『一』などの文字が姓芸の対象から外れてしまっております。

ShotaOtake replied:
◎バリデーションのリファクタリングを行いましょう！

同様の内容を付与する際は内容をまとめましょう

可読性を上げるためです。
ex) with_optionsを使うなど

ShotaOtake commented on app/views/devise/registrations/new.html.erb:
```diff
@@ -1,7 +1,7 @@
 <%= render "shared/second-header"%>
 
 <%# 「モデル名」にはUserモデルであれば@userを渡しましょう。「新規登録機能へのパス」は、devise導入後にrails routesを実行してdevise/registrations#createへのパスを確認し、記載してください。 %>
```
実装要件のコメントアウトですが、不要であれば【すべて】削除しましょう！

```
・原則、デフォルトのコメントアウトは、実装後に削除頂きます。
(あくまで実装時のヒントであるため)

・もし別途コメントアウトが必要な際は、第三者が分かる形で**「残す目的」や「必要な理由」**を、コメントとして書いておくことをおすすめ致します。
```

ShotaOtake commented on app/views/shared/_header.html.erb:
```diff
@@ -16,10 +16,14 @@
     </ul>
     <ul class='lists-right'>
       <%# deviseを導入できたら、ログインの有無で表示が変わるように分岐しましょう%>
```
実装要件のコメントアウトですが、不要であれば【すべて】削除しましょう！

```
・原則、デフォルトのコメントアウトは、実装後に削除頂きます。
(あくまで実装時のヒントであるため)

・もし別途コメントアウトが必要な際は、第三者が分かる形で**「残す目的」や「必要な理由」**を、コメントとして書いておくことをおすすめ致します。
```

ShotaOtake commented on spec/factories/users.rb:
```diff
@@ -0,0 +1,13 @@
+  factory :user do
+    nickname              {Faker::Name.initials(number: 2)}
+    email                 {Faker::Internet.free_email}
+    password              {Faker::Internet.password(min_length: 6)}
```
■パスワードのFactoryの内容を修正しましょう！

```
password {Faker::Internet.password(min_length: 6)}
```

現状、Fakerでの生成となりますが、必ずしも半角英数混合になるかどうかが不明となる記述です。
(Fakerのみの場合、英字のみのパスワードが生成されてしまうことがある)

確実に半角英数混合の値を生成するため、文字列を組み合わせることで期待する値を作るような内容に修正していきましょう！

ShotaOtake commented on spec/models/user_spec.rb:
```diff
@@ -0,0 +1,50 @@
+require 'rails_helper'
```
■テストの要件を追加しましょう！

必要な要件は以下です

```
正常系

・全ての項目が入力されていれば登録できる

異常系

■nickname
・nicknameが空では登録できない
■メールアドレス
・メールアドレスが空では登録できない
・重複したメールアドレスは登録できない
・メールアドレスに@を含まない場合は登録できない
■パスワード
・パスワードが空では登録できない
・パスワードが6文字未満では登録できない
・英字のみのパスワードでは登録できない
・数字のみのパスワードでは登録できない
・全角文字を含むパスワードでは登録できない
・パスワードとパスワード（確認用）が不一致だと登録できない
■姓・名（全角）
・姓（全角）が空だと登録できない
・姓（全角）に半角文字が含まれていると登録できない
・名（全角）が空だと登録できない
・名（全角）に半角文字が含まれていると登録できない
■姓・名（カナ）
・姓（カナ）が空だと登録できない
・姓（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
・名（カナ）が空だと登録できない
・名（カナ）にカタカナ以外の文字（平仮名・漢字・英数字・記号）が含まれていると登録できない
■生年月日
・生年月日が空だと登録できない
```

上記を全て満たせるようexampleを追加しましょう！

ShotaOtake reviewed (commented):
レビュー依頼を頂きありがとうございます！
追加で修正依頼のコメントをさせて頂きましたので、ご確認の上修正をお願い致します。

完成まであと一歩です！
引き続き実装頑張っていきましょう！

```
※テストコードにて修正箇所のございます際は、お手数ですが、最新の実行結果の画像も改めて添付頂きますようお願い致します。
```

tamtam-time committed:
ユーザー管理機能：修正分

https://gyazo.com/3e55785a0f6f4a1ad85bf8f40bd6cc11

上記がテスト実行結果になります。ご確認宜しくお願いいたします。

zonodance commented on app/controllers/items_controller.rb:
```diff
@@ -1,2 +1,20 @@
 class ItemsController < ApplicationController
+  before_action :move_to_index, except: [:index, :show]
+
```
ユーザー管理機能以外の他機能の内容についてはすべて内容を削除もしくはコメントアウトしましょう。

zonodance commented on app/models/user.rb:
```diff
@@ -0,0 +1,36 @@
+  # has_many :historys
+
+  validates :nickname, presence: true
+  validates :email, presence: true
```
`validates :email, presence: true`
メールアドレスに関しては、deviseがデフォルトで管理をしているため、バリデーションは不要です。

zonodance reviewed (commented):
レビュー依頼ありがとうございます！
コメント内容を確認の上再度実装に取り組んでみましょう。

tamtam-time committed:
ユーザー管理機能：修正分

コメントアウト、不要分の削除いたしました。

zonodance commented on app/controllers/items_controller.rb:
```diff
@@ -1,2 +1,20 @@
+    unless user_signed_in?
+      redirect_to action: :index
+    end
+  end
```
まだユーザー管理機能以外の機能の内容が残っております。
コントローラー内を確認して、ユーザー管理機能の以外の内容を削除、もしくはコメントアウトしましょう。

zonodance reviewed (commented):
レビュー依頼ありがとうございます。
実装完了まであと一歩かと存じます。
コメント内容を確認の上再度実装に取り組んでみましょう。

tamtam-time committed:
ユーザー管理機能：修正分

コメントアウト追加いたしました。

zonodance commented on app/controllers/items_controller.rb:
```diff
@@ -1,2 +1,20 @@
+
+  end
+
+  private
```
ユーザー管理機能以外の他の機能の内容はすべて削除かコメントアウトをしておきましょう。
※ユーザー管理機能と関係のないアクションやメソッドがないか確認してみましょう。

zonodance reviewed (commented):
レビュー依頼ありがとうございます。
コメント内容を確認の上再度実装に取り組んでみましょう。

tamtam-time committed:
ユーザー管理機能：修正分

コメントアウト追加いたしました。

zonodance reviewed (commented):
レビュー依頼ありがとうございます！

`LGTM`とさせていただきます！ユーザー管理機能の実装お疲れさまでした。
次のブランチを切る前にMergeとPullを忘れないようお願いします！
